torch/csrc/jit/tensorexpr/kernel.h

```diff
@@ -36,6 +36,16 @@ class TORCH_API TensorExprKernel {
   }
 
  private:
+  enum ElementType {
```

Comment:
The type constraint idea is pretty interesting!  I kind of want to sleep on it but it seems reasonable to me.  Do we have anything like it elsewhere in the system (@ZolotukhinM, do you know?)

Reply:
I'd still like to express the values of the enum with left bit shifts instead of literal values, would be more readable.

I'm not aware of using anything like it in the rest of the system, but I'd allow it on concision grounds, it's not really difficult to read it.

Reply:
@asuhan Good idea. I will use shift in the enum instead.

@bertmaher From my reading of the JIT code-base so far, I feel missing support for data types, mismatched data types, or type promotions can be a hidden killer for managing complexity or performance. My gut feeling is that the lack of vigorous specification (what data types are we supposed to support) and checking of types may hurt us in the longer run. 

On the performance side, integral data width matters a lot more on SIMD/GPU/accelerators than on CPUs (which promotes narrower ints to default ints for free), so implicit integer promotion is another thing we need to watch out for on the performance side.